### PR TITLE
Add incident column for when an incident occurred at

### LINKED ISSUE #2208: update_at value with dashboard
When I use the dashboard, created_at and updated_at are always equals.

My the typical workflow is:
Create an incident with the dashboard, and query the database: 

mysql> select * from incidents where id=2\G
*************************** 1. row ***************************
          id: 2
component_id: 0
        name: test_update_at
      status: 1
     visible: 1
     message: Test update_at
scheduled_at: NULL
  **created_at: 2016-10-24 14:09:15**
  **updated_at: 2016-10-24 14:09:15**
  deleted_at: NULL
1 row in set (0.00 sec)

Then I update the incident with the API (here I change the status): 
curl -X PUT -H "Content-Type: application/json" -H "X-Cachet-Token: xxxx" http://status.com/api/v1/incidents/2  -d '{"status":2}'

Now database have different values for created_at and updated_at: 
mysql> select * from incidents where id=2\G
*************************** 1. row ***************************
          id: 2
component_id: 0
        name: test_update_at
      status: 2
     visible: 1
     message: Test update_at
scheduled_at: NULL
  **created_at: 2016-10-24 14:09:15**
  **updated_at: 2016-10-24 14:14:08**
  deleted_at: NULL
1 row in set (0.00 sec)

Now if I edit the incident again with the dashboard an query the database: 

mysql> select * from incidents where id=2\G
*************************** 1. row ***************************
          id: 2
component_id: 0
        name: test_update_at
      status: 2
     visible: 1
     message: Test update_at
scheduled_at: NULL
  **created_at: 2016-10-24 14:09:00**
  **updated_at: 2016-10-24 14:09:00**
  deleted_at: NULL
1 row in set (0.01 sec)

It seems that the dashboard use the same value for created_at and updated_at.

### PR DESCRIPTION
**Pull request by jbrooksuk**
Closes #2208 

---

This does change the API requests, as we're now looking for a field called `occurred_at` instead of `incident_date`, or whatever.

### CODE REVIEW

File: app/Bus/Commands/Incident/ReportIncidentCommand.php [hunk 1]
```diff
@@ -110,7 +110,7 @@
         'component_status' => 'nullable|required_with:component_id|int|min:0|max:4',
         'notify'           => 'nullable|bool',
         'stickied'         => 'required|bool',
-        'incident_date'    => 'nullable|string',
+        'occurredAt'       => 'nullable|string',
```
Comment (ConnorVG) [inline app/Bus/Commands/Incident/ReportIncidentCommand.php]: I reckon we should move to `camelCase` but I also reckon it should be done in it's own PR rather than just leaving random bits here and there.

File: app/Bus/Handlers/Commands/Incident/UpdateIncidentCommandHandler.php [hunk 2]
```diff
@@ -61,18 +62,20 @@ public function handle(UpdateIncidentCommand $command)
...
-            $incidentDate = $this->dates->create('d/m/Y H:i', $command->incident_date);
-
-            $incident->update([
-                'created_at' => $incidentDate,
-                'updated_at' => $incidentDate,
-            ]);
+        if ($occurredAt = $command->occurredAt) {
+            if ($date = $this->dates->create('Y-m-d H:i', $occurredAt)) {
+                $incident->fill(['occurred_at' => $date]);
+            } else {
```
Comment (ConnorVG) [inline app/Bus/Handlers/Commands/Incident/UpdateIncidentCommandHandler.php]: Guard > `else` - looks messy imo

File: app/Http/Controllers/Dashboard/DashboardController.php [hunk 3]
```diff
@@ -127,11 +127,11 @@ public function showDashboard()
...
     protected function getIncidents()
     {
-        $allIncidents = Incident::notScheduled()->whereBetween('created_at', [
+        $allIncidents = Incident::notScheduled()->whereBetween('occurred_at', [
             $this->startDate->copy()->subDays(30)->format('Y-m-d').' 00:00:00',
             $this->startDate->format('Y-m-d').' 23:59:59',
-        ])->orderBy('created_at', 'desc')->get()->groupBy(function (Incident $incident) {
-            return (new Date($incident->created_at))
+        ])->orderBy('occurred_at', 'desc')->get()->groupBy(function (Incident $incident) {
+            return (new Date($incident->occurred_at))
```
Comment (ConnorVG) [inline app/Http/Controllers/Dashboard/DashboardController.php]: Could we not use the presenter here?

File: app/Http/Controllers/FeedController.php [hunk 4]
```diff
@@ -80,12 +85,12 @@ private function feedAction(ComponentGroup &$group, $isRss)
     {
         if ($group->exists) {
             $group->components->map(function ($component) use ($isRss) {
-                $component->incidents()->visible()->orderBy('created_at', 'desc')->get()->map(function ($incident) use ($isRss) {
+                $component->incidents()->visible()->orderBy('occurred_at', 'desc')->get()->map(function ($incident) use ($isRss) {
```
Comment (ConnorVG) [inline app/Http/Controllers/FeedController.php]: Long line is long :stuck_out_tongue:

File: app/Http/Controllers/StatusPageController.php [hunk 5]
```diff
@@ -109,7 +109,7 @@ public function showIndex()
             ->withDaysToShow($daysToShow)
             ->withAllIncidents($allIncidents)
             ->withCanPageForward((bool) $today->gt($startDate))
-            ->withCanPageBackward(Incident::notScheduled()->where('created_at', '<', $startDate->format('Y-m-d'))->count() > 0)
+            ->withCanPageBackward(Incident::notScheduled()->where('occurred_at', '<', $startDate->format('Y-m-d'))->count() > 0)
```
Comment (ConnorVG) [inline app/Http/Controllers/StatusPageController.php]: Do we even need to format `$startDate`?

File: app/Presenters/IncidentPresenter.php [hunk 6]
```diff
@@ -56,33 +56,73 @@ public function raw_message()
     }
 
     /**
+     * Present formatted occurred_at date time.
+     *
+     * @return string
+     */
+    public function occurred_at()
+    {
+        return app(DateFactory::class)->make($this->wrappedObject->occurred_at)->toDateTimeString();
```
Comment (ConnorVG) [inline app/Presenters/IncidentPresenter.php]: Off topic but is there any reason we don't just inject the DateFactory?

File: app/Presenters/IncidentPresenter.php [hunk 7]
```diff
@@ -56,33 +56,73 @@ public function raw_message()
...
     /**
      * Present formatted date time.
      *
      * @return string
      */
-    public function created_at_formatted()
+    public function occurred_at_formatted()
     {
-        return ucfirst(app(DateFactory::class)->make($this->wrappedObject->created_at)->format(Config::get('setting.incident_date_format', 'l jS F Y H:i:s')));
+        return ucfirst(app(DateFactory::class)->make($this->wrappedObject->occurred_at)->format(Config::get('setting.incident_date_format', 'l jS F Y H:i:s')));
```
Comment (ConnorVG) [inline app/Presenters/IncidentPresenter.php]: This seems overly complex / long. QoL goes down.

File: app/Presenters/IncidentPresenter.php [hunk 8]
```diff
@@ -56,33 +56,73 @@ public function raw_message()
...
+    }
+
+    /**
+     * Present formatted date time.
+     *
+     * @return string
+     */
+    public function created_at_formatted()
+    {
+        return ucfirst(app(DateFactory::class)->make($this->wrappedObject->created_at)->format(Config::get('setting.incident_date_format', 'l jS F Y H:i:s')));
```
Comment (ConnorVG) [inline app/Presenters/IncidentPresenter.php]: Same as previous.

File: app/helpers.php [hunk 9]
```diff
@@ -177,3 +177,61 @@ function cachet_redirect($name, $parameters = [], $status = 302, $headers = [],
...
+     *
+     * Thanks to http://stackoverflow.com/a/30192680/394013
+     *
+     * @param string $format
+     *
+     * @return string
+     */
+    function datetime_to_moment($format)
+    {
+        $replacements = [
```
Comment (ConnorVG) [inline app/helpers.php]: This is just bizarre. I see it's point tho :laughing:

File: resources/views/single-incident.blade.php [hunk 10]
```diff
@@ -7,7 +7,7 @@
 @stop
 
 @section('content')
-<h1>{{ $incident->name }} <small>{{ formatted_date($incident->created_at) }}</small></h1>
+<h1>{{ $incident->name }} <small>{{ formatted_date($incident->occurred_at) }}</small></h1>
```
Comment (ConnorVG) [inline resources/views/single-incident.blade.php]: Why are we not using the presenter here?

File: app/Bus/Handlers/Commands/Incident/UpdateIncidentCommandHandler.php (hunk 2 above)
Comment (jbrooksuk) [inline app/Bus/Handlers/Commands/Incident/UpdateIncidentCommandHandler.php]: Also, the exception doesn't do anything here, because `create` will throw, so we need to catch.

File: app/Http/Controllers/Dashboard/DashboardController.php (hunk 3 above)
Comment (jbrooksuk) [inline app/Http/Controllers/Dashboard/DashboardController.php]: I was thinking this when I was looking at it this morning.

File: app/Http/Controllers/FeedController.php (hunk 4 above)
Comment (jbrooksuk) [inline app/Http/Controllers/FeedController.php]: I should add a better scope for this.

File: app/Http/Controllers/StatusPageController.php (hunk 5 above)
Comment (jbrooksuk) [inline app/Http/Controllers/StatusPageController.php]: Dunno.

File: app/Presenters/IncidentPresenter.php (hunk 6 above)
Comment (jbrooksuk) [inline app/Presenters/IncidentPresenter.php]: Laziness.

File: app/Presenters/IncidentPresenter.php (hunk 7 above)
Comment (jbrooksuk) [inline app/Presenters/IncidentPresenter.php]: :+1:

File: app/helpers.php (hunk 9 above)
Comment (jbrooksuk) [inline app/helpers.php]: I didn't end up using this, but I like it, so I'm keeping it because I probably will use this at some point.

File: resources/views/single-incident.blade.php (hunk 10 above)
Comment (jbrooksuk) [inline resources/views/single-incident.blade.php]: I noticed this too, dunno why.

File: app/Http/Controllers/StatusPageController.php (hunk 5 above)
Comment (jbrooksuk) [inline app/Http/Controllers/StatusPageController.php]: Yes, because `occurred_at` might contain a time, but we're only interested in whole days.

**Comment by jbrooksuk** [comment]
Let's merge and see what happens 👍 there might be an edge-case I've missed.
